File: algorithms/code/leetcode/lc199_binary_tree_right_side_view/lc199_binary_tree_right_side_view.py
# https://leetcode.com/problems/binary-tree-right-side-view/

from __future__ import annotations

import unittest


class TreeNode:
    def __init__(
        self,
        val: int = 0,
        left: 'TreeNode | None' = None,
        right: 'TreeNode | None' = None,
    ) -> None:
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def __rightSideView(
        self, root: TreeNode | None, level: int, listed: list[int]
    ) -> None:
        if root is not None:
            level += 1
            if level > len(listed):
                listed.append(root.val)
            self.__rightSideView(root.right, level, listed)
            self.__rightSideView(root.left, level, listed)

    def rightSideView(self, root: TreeNode | None) -> list[int]:
        listed: list[int] = []
        self.__rightSideView(root, 0, listed)
        return listed


class TestCode(unittest.TestCase):
    def test_example(self) -> None:
        n1 = TreeNode(1)
        n2 = TreeNode(2)
        n3 = TreeNode(3)
        n4 = TreeNode(4)
        n5 = TreeNode(5)
        n6 = TreeNode(6)
        n1.left = n2
        n1.right = n3
        n2.left = None
        n2.right = n5
        n3.left = None
        n3.right = n4
        n4.left = None
        n4.right = None
        n5.left = n6
        n5.right = None
        n6.left = None
        n6.right = None
        expected = [1, 3, 4, 6]
        result = Solution().rightSideView(n1)
        for i, expected_i in enumerate(expected):
            self.assertEqual(expected_i, result[i])

File: algorithms/code/leetcode/lc199_binary_tree_right_side_view/lc199_binary_tree_right_side_view.rs
// https://leetcode.com/problems/binary-tree-right-side-view/
// #easy

use std::cell::RefCell;
use std::rc::Rc;

#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    pub fn new(
        val: i32,
        left: Option<Rc<RefCell<TreeNode>>>,
        right: Option<Rc<RefCell<TreeNode>>>,
    ) -> Option<Rc<RefCell<Self>>> {
        Some(Rc::new(RefCell::new(TreeNode { val, left, right })))
    }
}

pub struct Solution;

impl Solution {
    fn right_side_view_helper(
        root: Option<Rc<RefCell<TreeNode>>>,
        level_value: usize,
        list: &mut Vec<i32>,
    ) {
        if let Some(root) = root {
            let root = root.borrow();
            let level = level_value + 1;
            if level > list.len() {
                list.push(root.val);
            }
            Self::right_side_view_helper(root.right.clone(), level, list);
            Self::right_side_view_helper(root.left.clone(), level, list);
        }
    }

    pub fn right_side_view(root: Option<Rc<RefCell<TreeNode>>>) -> Vec<i32> {
        let mut list = Vec::new();
        Self::right_side_view_helper(root, 0, &mut list);
        list
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_example() {
        let n6 = TreeNode::new(6, None, None);
        let n5 = TreeNode::new(5, n6, None);
        let n4 = TreeNode::new(4, None, None);
        let n3 = TreeNode::new(3, None, n4);
        let n2 = TreeNode::new(2, None, n5);
        let n1 = TreeNode::new(1, n2, n3);
        let result = Solution::right_side_view(n1);
        let expected = vec![1, 3, 4, 6];
        assert_eq!(result, expected);
    }
}
